pkg/codegen/importer/hcl2.go

```diff
@@ -0,0 +1,467 @@
+	// TODO: pull the package version from the resource's provider
+	pkg, err := loader.LoadPackage(string(state.Type.Package()), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	r, ok := pkg.GetResource(string(state.Type))
+	if !ok {
+		return nil, fmt.Errorf("unknown resource type '%v'", r)
+	}
+
+	var items []model.BodyItem
+	for _, p := range r.InputProperties {
+		x, err := generatePropertyValue(p, state.Inputs[resource.PropertyKey(p.Name)])
+		if err != nil {
+			return nil, err
+		}
+		if x != nil {
+			items = append(items, &model.Attribute{
+				Name:  p.Name,
+				Value: x,
+			})
+		}
+	}
+
+	resourceOptions, err := makeResourceOptions(state, names)
+	if err != nil {
+		return nil, err
+	}
+	if resourceOptions != nil {
+		items = append(items, resourceOptions)
+	}
+
+	typ, name := state.URN.Type(), state.URN.Name()
+	return &model.Block{
+		Tokens: syntax.NewBlockTokens("resource", string(name), string(typ)),
+		Type:   "resource",
+		Labels: []string{string(name), string(typ)},
+		Body: &model.Body{
+			Items: items,
+		},
+	}, nil
+}
+
+func newVariableReference(name string) model.Expression {
+	return model.VariableReference(&model.Variable{
+		Name:         name,
+		VariableType: model.DynamicType,
+	})
+}
+
+func appendResourceOption(block *model.Block, name string, value model.Expression) *model.Block {
+	if block == nil {
+		block = &model.Block{
+			Tokens: syntax.NewBlockTokens("options"),
+			Type:   "options",
+			Body:   &model.Body{},
+		}
+	}
+	block.Body.Items = append(block.Body.Items, &model.Attribute{
+		Tokens: syntax.NewAttributeTokens(name),
+		Name:   name,
+		Value:  value,
+	})
+	return block
+}
+
+func makeResourceOptions(state *resource.State, names NameTable) (*model.Block, error) {
+	var resourceOptions *model.Block
+	if state.Parent != "" && state.Parent.Type() != resource.RootStackType {
+		name, ok := names[state.Parent]
+		if !ok {
+			return nil, fmt.Errorf("no name for parent %v", state.Parent)
+		}
+		resourceOptions = appendResourceOption(resourceOptions, "parent", newVariableReference(name))
+	}
+	if state.Provider != "" {
+		ref, err := providers.ParseReference(state.Provider)
+		if err != nil {
+			return nil, fmt.Errorf("invalid provider reference %v: %w", state.Provider, err)
+		}
+		if !providers.IsDefaultProvider(ref.URN()) {
+			name, ok := names[ref.URN()]
+			if !ok {
+				return nil, fmt.Errorf("no name for provider %v", state.Parent)
+			}
+			resourceOptions = appendResourceOption(resourceOptions, "provider", newVariableReference(name))
+		}
+	}
+	if len(state.Dependencies) != 0 {
+		deps := make([]model.Expression, len(state.Dependencies))
+		for i, d := range state.Dependencies {
+			name, ok := names[d]
+			if !ok {
+				return nil, fmt.Errorf("no name for resource %v", d)
+			}
+			deps[i] = newVariableReference(name)
+		}
+		resourceOptions = appendResourceOption(resourceOptions, "dependsOn", &model.TupleConsExpression{
```

Comment:
Might be nice to create named types for these resource options (`parent`, `provider`, etc)